# Reversal Lab: Reversal Competition
### Data Science for Biology
Developed by Steven E. Brenner

## Instructions

Provide blocks of code that take a string 's' and put the reverse of that string into 'r'.
For each one, print out 's' and 'r' unless instructed otherwise.

The primary evaluation will be on which team comes up with the largest number of distinct ways of reversing the string, which should be listed in as many code blocks as needed at the bottom of the assignment (under "count the reversals"). Feel free to use that space to experiment.

Fill in the 3 boxes with the rapid, concise, and (optionally) obscure examples for the additional competitions.

Bonus prizes may be awarded for best team name, as well as the fastest, shortest, and most obscure examples. Winners get bragging rights for perpetuity. 

You will be graded only on completion for this lab.

## Group Team **[List Team Name Here]**
This should be a creative name for your team.

Team members: **[List members here]**


### Fastest code 

Use the code framework and example string below to test the speed of your fastest code.  

Note that `%%timeit` counts the time for the entire cell, including assigning the string to `s`

Therefore, we first figure out how long it takes to just do that assignment, and then we
calculate the time to put the reverse of that string into `r`.

Here do NOT print the sequences, as those could dominate the timing.

Note: Repeat your code from "all examples" below as part of the regular set of examples to have it counted. <span style="color: red;">

Our string for this one will be a lowercase, punctuation- and space-free version of **`"Madam, in Eden, I'm Adam."`**
Everyone must use the same string, so timing will be consistent.  Then copy and paste the timings you had for each into the box below

In [1]:
%%timeit -n100000 -r 100 

s = "madaminedenimadam"

# Don't change this code block!
# This will be the speed of just doing the string assignment and reading this comment

6.86 ns ± 0.842 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)


In [4]:
%%timeit -n100000 -r 100 

s = "madaminedenimadam"

r = s[::-1]

# Change this code block!
# This will be the speed of the string assignment, reversal, and reading this comment

57.6 ns ± 3.03 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)


#### Code speed 

Copy into this box the two lines of output from the code boxes above, and then list the difference, as that will be roughly the runtime of your reversing code.  **Replace the lines below with the output that you got, and edit difference between them**

It should look something like:

19 ns ± 4.88 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)

171 ns ± 38.7 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)

Difference in runtime = **152ns**

#### Your code speed

6.86 ns ± 0.842 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)

57.6 ns ± 3.03 ns per loop (mean ± std. dev. of 100 runs, 100,000 loops each)

Difference in runtime = **50.74ns**

### Concise code

Enter your code with the fewest characters below to reverse the string successfully.  

Precisely repeat your code example below as part of the regular set of "all examples", including printing 
the original and reversed sequence.  Use any string you like, although one is provided to you by default..

In [3]:
s = "You lose"  # Concise quip from laconic President Calvin Coolidge

# Your reversing code here 
r = s[::-1]

print ("String  :",s,"\nReversed:",r)

String  : You lose 
Reversed: esol uoY


### Obscure code

Enter your code that most effectively obscures that r will be assigned the reverse of string s.  

This need not be one of your examples below (it may just use an obscured version of those). 

If is is the same as one of your examples below, precisely repeat your code example here.

Print the original and reversed sequence.  Do not change the original sequence s

In [ ]:
clergy_concerns = "dennis nell edna leon nedra anita rolf nora alice carol leo jane reed denadale dasil rae penny lana dave denny lena ida bernadette ben ray lila nina jo ira mara sara mario jan ina lily arne bette dan reba diane lynn ed eva dana lynne pearl isabel ada ned dee rena joel lora secil aaron flora tina arden noel and ellen sinned"
s = clergy_concerns.replace(" ", "")

# Your reversing code here

print ("String  :",s,"\nReversed:",r)


## Count the reversals

Have one code block for each reversal. Add additional code boxes as needed.

Arrange the most similar approaches together

You can pick the phrase to reverse.

Feel free to add comments about your approach (why you like it, how you came up with it, etc.)

In [7]:
# Approach 1
s = "It is a tale / Told by an idiot, full of sound and fury, / Signifying nothing."

# Your reversing code here
def reverse_string(s):
    if len(s) == 0:
        return ""
    temp = s[0]
    return reverse_string(s[1:]) + temp
    
r = reverse_string(s)

print ("String  :",s,"\nReversed:",r)


String  : It is a tale / Told by an idiot, full of sound and fury, / Signifying nothing. 
Reversed: .gnihton gniyfingiS / ,yruf dna dnuos fo lluf ,toidi na yb dloT / elat a si tI


In [9]:
# Approach 2
import collections

s = "Definition of a college professor: someone who talks in other people's sleep." # W.H. Auden

# Your reversing code here
# We create a linked List
linked_list = collections.deque()

# use doubled ended queue and appendLeft to reverse 0_0
for char in s:
    linked_list.appendleft(char)

r = ''.join(linked_list)

print ("String  :",s,"\nReversed:",r)


String  : Definition of a college professor: someone who talks in other people's sleep. 
Reversed: .peels s'elpoep rehto ni sklat ohw enoemos :rosseforp egelloc a fo noitinifeD


In [16]:
# Approach 3
s = "Let's drink a toast as each of us recalls / Ivy-covered professors in ivy-covered halls" # Tom Lehrer, "Bright College Days"

# Your reversing code here
linked_list = collections.deque()

for char in s:
    linked_list.append(char)

# Reverse the deque
linked_list.reverse()

r = ''.join(linked_list)

print ("String  :",s,"\nReversed:",r)


String  : Let's drink a toast as each of us recalls / Ivy-covered professors in ivy-covered halls 
Reversed: sllah derevoc-yvi ni srosseforp derevoc-yvI / sllacer su fo hcae sa tsaot a knird s'teL


In [28]:
from transformers import BartForConditionalGeneration, BartTokenizer

# Load pre-trained model and tokenizer
model_name = 'facebook/bart-large'
model = BartForConditionalGeneration.from_pretrained(model_name)
tokenizer = BartTokenizer.from_pretrained(model_name)

def reverse_string_with_llm(s):
    # Prepare the input prompt
    prompt = f"output the reverse of the string: {s}"

    # Encode the input prompt
    inputs = tokenizer(prompt, return_tensors='pt')

    # Generate the output
    outputs = model.generate(inputs['input_ids'], max_length=100, num_return_sequences=1)
    print(outputs)

    # Decode the output
    reversed_string = tokenizer.decode(outputs[0], skip_special_tokens=True)
    print(reversed_string)
    # Extract the reversed string from the output
    reversed_string = reversed_string.split(':')[-1].strip()

    return reversed_string

# Example usage
s = "hola como estas"
r = reverse_string_with_llm(s)

print("String  :", s)
print("Reversed:", r)

tensor([[    2,     0, 46234,     5,  7213,     9,     5,  6755,    35,  1368,
          3019,  3137,   139,  3304,   281,     2]])
output the reverse of the string: hola como estas
String  : hola como estas
Reversed: hola como estas
